### ISSUE TITLE
Add job descriptions in the API

### ISSUE THREAD

**Issue by iedmrc** [VROOM-Project/vroom#235]
vroom-frontend supports job descriptions and shows on the map as well, but when requesting to vroom through express.js API, job descriptions are lost in response. Do vroom-backend & vroom express.js API support job descriptions natively ? If not, do you consider adding job descriptions natively in vroom ? Hence, we can throw the burden of extra scripting away. (I mean, I have 500 jobs and 30 vehicles for each day of the week, so iterating over such a big array sounds not good 🙂).

Thanks!

**Comment by jcoupey** [comment]
You're right, the description part is currently only handled on the frontend side, and the `description` keys are retrieved from the input jobs [here](https://github.com/VROOM-Project/vroom-frontend/blob/7814c0737a5379d2d47e3b44296110ba52117197/src/utils/data_handler.js#L748), not from the solution itself. To avoid the extra iterations you mention, we store job ranks based on their id [in one go](https://github.com/VROOM-Project/vroom-frontend/blob/7814c0737a5379d2d47e3b44296110ba52117197/src/utils/data_handler.js#L713-L716).

This is clearly not ideal and adding a `description` field for jobs in the API could make sense. Let me know if you want to work on this feature, I can provide a few pointers.

One thing we'd have to figure out is how we handle the same for vehicles. A description for a job or its location is pretty much the same, but a vehicle can have a start and end. So we'd have to decide if we want a description related to the vehicle or to its locations (or both).

**Comment by iedmrc** [comment]
> This is clearly not ideal and adding a description field for jobs in the API could make sense. Let me know if you want to work on this feature, I can provide a few pointers.

I would love to work on this feature, but I really don't have enough knowledge of c++ to ship the feature. I just can read and fix little bugs in c++ :)

> One thing we'd have to figure out is how we handle the same for vehicles. A description for a job or its location is pretty much the same, but a vehicle can have a start and end. So we'd have to decide if we want a description related to the vehicle or to its locations (or both).

To me, vehicles should have a description associated with itself, not the locations. But you have more comprehensive knowledge of the project than me.

**Comment by jcoupey** [comment]
> I really don't have enough knowledge of c++

Not such a big deal actually: the required steps are to parse an additional field from the input, store it in a new member variable for job objects, and then write it back in output when formatting the route steps.

>  vehicles should have a description associated with itself

I guess both might be useful. Currently the frontend does use a description for start and end to display adresses.

**Comment by iedmrc** [comment]
I am not sure but if I can/do before you or someone, I will send a pull request. Thanks!

**Comment by tracify08** [comment]
Hi Julien,

Any hope of this feature being implemented? I think it will add a lot more value. 

Thank you

**Comment by jcoupey** [comment]
Yes, this has been kind of laid aside because there is no added optimization functionality, but I agree it would be convenient to add. I'm flagging this for the next milestone.

One additional question (on top of the vehicle thing above) is to know where to put the description on a shipment. It could be either at `shipment` level or at both `shipment.pickup` and `shipment.delivery` level, which would enable using different descriptions.

In other words, do we describe a `job`, `shipment` or `vehicle`, or do we describe the associated locations. I'd be inclined to do the later as it is more flexible and above all it would be consistent with how the description would appear in output (at `step` level).

**Comment by jcoupey** [comment]
For consistency (and because it could actually be useful), we should also have break descriptions.

**Comment by jcoupey** [comment]
Somehow merging #356 did not close this issue automatically due to a branch rebase and force push. But this landed in b5dd59d and we now have description for `vehicle`, `job`, `pickup`, `delivery` and `break` objects.
